docs/src/references/mqtt-topic-guidelines.md

```diff
@@ -0,0 +1,608 @@
+
+
+**Why have the `/service/` subtopic level and not have the <service-id> directly?**
+
+Primarily for future-proofing as `service` is a new kind of child abstraction that we've added now.
+If another abstraction is introduced in the future, say `plugin`, they can be namespaced under a `plugin/` subtopic.
+It also makes the distinction from `device` data clearer.
+
+### Telemetry
+
+For telemetry data, the topics would be grouped under a `telemetry/` sub-topic as follows:
+
+|Type|Example|
+|------|-------|
+|Measurements|`<target_prefix>/telemetry/measurements`|
+|Events|`<target_prefix>/telemetry/events/<event-type>`|
+|Alarms|`<target_prefix>/telemetry/alarms/<alarm-type>/<severity>`|
+
+**Examples**
+
+The following shows an example of the measurement telemetry topics for each of the target types:
+
+|Target|Topic structure|
+|------|-------|
+|tedge device|`tedge/main/device/telemetry/measurements`|
+|service on tedge device|`tedge/main/service/<service-id>/telemetry/measurements`|
+|child device|`tedge/<child-id>/device/telemetry/measurements`|
+|service on child device|`tedge/<child-id>/service/<service-id>/telemetry/measurements`|
+
+**Why have the redundant `/telemetry` subtopic level?**
+
+To have a clear separation from commands or other kinds of data that might get added in future.
+It simplifies subscriptions for "all telemetry data from a device" to just `tedge/<device-id>/+/telemetry/#`.
+Without that `telemetry` grouping level, doing such a subscription would be difficult as you'll have to subscribe to each of the following topics separately:
+
+* `tedge/<device-id>/+/measurements/#`
+* `tedge/<device-id>/+/events/#`
+* `tedge/<device-id>/+/alarms/#`
+
+In additional, this topic structure allow to combine the subscription of telemetry data as well as command data via the following MQTT wildcard topic:
+
+```
+tedge/<device-id>/+/#
+```
+
+Creating ACL rules is also simplified by allowing telemetry data to be controlled independently to the command topics.
+
+### Commands
+
+Similarly, all commands would be grouped under a `commands/` sub-topic as follows:
+
+|Target|Topic structure|
+|------|-------|
+|For requests|`tedge/<device-id>/<target-type>/commands/req/<operation-type>/<operation-specific-keys>...`|
+|For responses|`tedge/<device-id>/<target-type>/commands/res/<operation-type>/<operation-specific-keys>...`|
+
+The `operation-specific-keys` are optional and the number of such keys (topic levels) could vary from one operation to another.
+
+Child devices follow a similar structure for commands as well:
+
+```
+tedge/<child-id>/commands/req/software_list
+tedge/<child-id>/commands/res/software_update
+```
+
+
+If command support is added to services in future, they'd follow a similar pattern as in:
+
+```
+tedge/<device-id>/service/<service-id>/commands/req/software_list
+tedge/<device-id>/service/<service-id>/commands/req/software_list
+```
```

Comment:
`restart` would be a more appropriate example.

Reply:
Though shouldn't the examples be more inline with the fore-mentioned schema, e.g. `software/list` instead of `software_list`?

Full example:

```sh
tedge/<device-id>/service/<service-id>/commands/req/software/list
tedge/<device-id>/service/<service-id>/commands/req/software/update
```

Reply:
As discussed, not adding the operation type grouping level, for little benefits.